server/neptune/workflows/internal/deploy/revision/queue/worker.go

```diff
@@ -100,25 +100,22 @@ func (w *Worker) Work(ctx workflow.Context) {
 
 		// This should only happen on startup
 		// If we fail to fetch the latest deployment, log and exit the workflow
-		if w.LatestDeployment == nil {
-			w.LatestDeployment, err = w.fetchLatestDeployment(ctx, msg)
+		if latestDeployment == nil {
+			latestDeployment, err = w.fetchLatestDeployment(ctx, msg)
 			if err != nil {
 				logger.Error(ctx, fmt.Sprint("Unable to fetch latest deployment, worker is shutting down", err.Error()))
 				return
 			}
 		}
 
-		var compareCommitResp activities.CompareCommitResponse
-		err = workflow.ExecuteActivity(ctx, w.Activities.CompareCommit, activities.CompareCommitRequest{
-			DeployRequestRevision:  msg.Revision,
-			LatestDeployedRevision: w.LatestDeployment.Revision,
-			Repo:                   msg.Repo,
-		}).Get(ctx, &compareCommitResp)
+		shouldDeployRevision, err := w.proecessRevision(ctx, msg, latestDeployment)
```

Comment:
This is the basically the same thing as before, you're doing a bunch of the logic in the function to sift through something that can have more than two states and then returning a variable that only has 2 states.  This doesn't really help when adding force applies and will require some refactor.


Instead i would probably do something like this:

```go

func Work() {

   var lastDeployment 
   for {
        awaitQueueNotEmpty()
        msg :=  queue.Pop()
         
        args := validateAndBuildArgs(msg)
        
        err := w.TerraformWorkflowRunner.Run(ctx, args)

        persistNewDeployment(args.DeploymentInfo)
        
        lastDeployment = args.DeploymentInfo

    }
}

func validateAndBuildArgs(lastDeployment, msg) SomeType {

    var args SomeType
    if lastDeployment != nil {
        fetchLatestDeployment(...)
    }

    comparison, err := getCommitComparison(...)

    switch comparison {
         case behind:
         case ahead:
         // TODO: Support force applies by adding a new field on args
         case diverged:
         ....

    }

    return args
}
```

Reply:
From my understanding and checking with @samrabelachew , we might not need to change the `TerraformWorkflowRunner` method for supporting FA since the the workflow runner does not need to know about the root locking mechanism. So, instead of building args for the workflow runner, I decided to refactor the existing implementation by adding a check for invalid revision before executing the `processRevision` method to only handle valid revisions. This eliminates the problem of having to return two state variables in the `procesRevision` method. instead, it only returns an error.

Reply:
My PR is just a draft, but what are your thoughts on just consolidating all of the logic to the `processRevision` like so: https://github.com/lyft/atlantis/blob/a0c72a0f4113757701bdb9dab60338107a5449c0/server/neptune/workflows/internal/deploy/revision/queue/worker.go#L111. Maybe even isolating that code out into a separate `RevisionProcessor`?

Reply:
separating it out sounds fine with me, but since this is fixes a nil pointer let's just get this out for now

Reply:
The issue with returning only an error is we have to determine whether to run the workflow or not based on the error. So, we have to return an error for `DirectionBehind` commits to tell the calling function to skip running the tf workflow. When we add FA logic in here, it'll be even more confusing since the error could be from FA failure too. 

Separating it out sounds fine to me as well.